### PR TITLE
Add `InputEvent`, `StaticRange`, `AbstractRange`

### PR DESCRIPTION
**Pull request by armanbilge**
Just wanted to add `InputEvent` and it sort of snowballed :)

- https://developer.mozilla.org/en-US/docs/Web/API/AbstractRange
- https://developer.mozilla.org/en-US/docs/Web/API/StaticRange
- https://developer.mozilla.org/en-US/docs/Web/API/InputEvent
- ~https://www.w3.org/TR/input-events-2/#interface-InputEvent-Attributes~
  MDN links to this, which is different (?)
  https://rawgit.com/w3c/input-events/v1/index.html#interface-InputEvent-Attributes

### CODE REVIEW

File: dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  /** The offset into the node indicated by endOffset at which the last character in the range is located. */
+  val endOffset: Int
+
+  /** The Node in which the starting point of the range is located. */
+  val startContainer: Node
+
+  /** The Node in which the end point of the range is located. */
+  val endContainer: Node
+}
```
Comment (armanbilge) [inline dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala:18]: This is an unusual init, since none of the parameters are optional.

**Comment by armanbilge** [comment]
After we land this one I'm hoping we can cut a v2.2.0, now that we've accumulated a handful of contributor PRs :)

File: dom/src/main/scala/org/scalajs/dom/InputEventInit.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+package org.scalajs.dom
+
+import scala.scalajs.js
+
+trait InputEventInit extends UIEventInit {
+  var inputType: js.UndefOr[InputType] = js.undefined
+  var data: js.UndefOr[String] = js.undefined
+  var dataTransfer: js.UndefOr[DataTransfer] = js.undefined
+  var isComposing: js.UndefOr[Boolean] = js.undefined
+  var ranges: js.UndefOr[js.Array[StaticRange]]
```
Comment (armanbilge) [inline dom/src/main/scala/org/scalajs/dom/InputEventInit.scala:16]:
```suggestion
  var ranges: js.UndefOr[js.Array[StaticRange]] = js.undefined
```

**Comment by armanbilge** [comment]
Polite ping @japgolly :) if you don't have time to review for a while, no big deal, just let me know and I won't block v2.2.0 on this if that's okay with you!

**Comment by japgolly** [comment]
> Polite ping @japgolly :) if you don't have time to review for a while, no big deal, just let me know and I won't block v2.2.0 on this if that's okay with you!

Oh my! Apologies @armanbilge ! I haven't been keeping up with my email lately so I missed this. I'll get to this very son

**Comment by armanbilge** [comment]
No worries, whenever you get to it! I hope you don't mind I released v2.2.0, it was really wonderful to get a bunch of contributions and I wanted to get it out :)

File: dom/src/main/scala/org/scalajs/dom/InputEventInit.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  * Everything else is under the MIT License http://opensource.org/licenses/MIT
+  */
+package org.scalajs.dom
+
+import scala.scalajs.js
+
+trait InputEventInit extends UIEventInit {
+  var inputType: js.UndefOr[InputType] = js.undefined
+  var data: js.UndefOr[String] = js.undefined
+  var dataTransfer: js.UndefOr[DataTransfer] = js.undefined
```
Comment (japgolly) [inline dom/src/main/scala/org/scalajs/dom/InputEventInit.scala:14]: `dataTransfer` not allowed here

File: dom/src/main/scala/org/scalajs/dom/InputEventInit.scala [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  var isComposing: js.UndefOr[Boolean] = js.undefined
+  var ranges: js.UndefOr[js.Array[StaticRange]] = js.undefined
```
Comment (japgolly) [inline dom/src/main/scala/org/scalajs/dom/InputEventInit.scala:16]: `ranges` not allowed here

File: dom/src/main/scala/org/scalajs/dom/StaticRange.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/** All documentation for facades is thanks to Mozilla Contributors at https://developer.mozilla.org/en-US/docs/Web/API
+  * and available under the Creative Commons Attribution-ShareAlike v2.5 or later.
+  * http://creativecommons.org/licenses/by-sa/2.5/
+  *
+  * Everything else is under the MIT License http://opensource.org/licenses/MIT
+  */
+package org.scalajs.dom
```
Comment (japgolly) [inline dom/src/main/scala/org/scalajs/dom/StaticRange.scala:7]: unnecessary (and false) package comment

File: dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala (hunk 1 above)
Comment (japgolly) [reply dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala:18]: Should we provide `object StaticRangeInit { def apply(...) = ... }` too?

**Review by japgolly — CHANGES_REQUESTED** [review]

File: dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala (hunk 1 above)
Comment (armanbilge) [reply dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala:18]: Good question. Seems reasonable, but I'm not sure if it's consistent with our style. I haven't gone through them carefully, but I think some of the crypto init-style objects may be a mix of required and optional parameters. Should we add `apply` methods for those too?

Comment (japgolly) [reply dom/src/main/scala/org/scalajs/dom/StaticRangeInit.scala:18]: Ah, good point. I'll leave that question open and consider it out of scope for this PR then.

**Review by japgolly — APPROVED** [review]
LGTM. Thanks for the follow-up PR and your patience :smile:

**Pull request merged**